components/sup/src/event/nats_message_stream.rs

```diff
@@ -4,27 +4,96 @@ use crate::event::{Error,
 use futures::{channel::{mpsc as futures_mpsc,
                         mpsc::UnboundedSender},
               stream::StreamExt};
-use rants::{error::Error as RantsError,
-            native_tls::TlsConnector,
-            Client,
-            Subject};
-use tokio::time;
+use nats::{self,
+           asynk::{Connection,
+                   Options}};
+use std::{path::PathBuf,
+          sync::Arc,
+          thread,
+          time::Duration};
+use tokio::{sync::Mutex,
+            time};
 
 /// The subject and payload of a NATS message.
 #[derive(Debug)]
 pub struct NatsMessage {
-    subject: &'static Subject,
+    subject: &'static str,
     payload: Vec<u8>,
 }
 
 impl NatsMessage {
-    pub fn new(subject: &'static Subject, payload: Vec<u8>) -> Self {
-        NatsMessage { subject, payload }
-    }
+    pub fn new(subject: &'static str, payload: Vec<u8>) -> Self { NatsMessage { subject, payload } }
 
     pub fn payload(&self) -> &[u8] { self.payload.as_slice() }
 }
 
+/// NatsClient is main accessor to connect to NATS Server and
+/// to publish messages to NATS.
+#[derive(Clone)]
+struct NatsClient(Arc<Mutex<Option<Connection>>>);
+
+impl NatsClient {
+    fn new() -> NatsClient { NatsClient(Arc::new(Mutex::new(None))) }
+
+    // Connect to the server. If a timeout was set, we want to ensure we establish a connection
+    // before exiting the function. If we do not connect within the timeout we return an error.
+    // If we do not have a timeout, we don't care if we can immediately connect. Instead we spawn
+    // a future that will resolve when a connection is possible. Once we establish a
+    // connection, this client will handle reconnecting if necessary.
+    async fn connect(self, supervisor_id: String, config: EventStreamConfig) -> Result<()> {
+        match config.connect_method.into() {
+            Some(timeout) => {
+                time::timeout(timeout, self.connect_impl(supervisor_id, &config)).await??;
+            }
+            None => {
+                tokio::spawn(async move { self.connect_impl(supervisor_id, &config).await });
+            }
+        }
+        Ok(())
+    }
+
+    async fn connect_impl(self, supervisor_id: String, config: &EventStreamConfig) -> Result<()> {
+        while self.0.lock().await.is_none() {
+            match Self::options_from_config(&supervisor_id, config)?.connect(&config.url
+                                                                                    .to_string())
+                                                                    .await
+            {
+                Ok(conn) => *self.0.lock().await = Some(conn),
+                Err(e) => {
+                    trace!("Failed to connect to NATS server: {}", e);
```

Comment:
I ended up having issues getting this to render and had to use `println!`. I think this might have something to do with the rust `log` crate marcros being used in a separate thread but will need some research.